src/array_api_extra/_funcs.py

```diff
@@ -548,3 +559,279 @@ def sinc(x: Array, /, *, xp: ModuleType | None = None) -> Array:
+            beneficial for performance.
+            You should not reuse it after calling this function.
+    xp : array_namespace, optional
+        The standard-compatible namespace for `x`. Default: infer
+
+    **kwargs:
+        If the backend supports an `at` method, any additional keyword
+        arguments are passed to it verbatim; e.g. this allows passing
+        ``indices_are_sorted=True`` to JAX.
+
+    Returns
+    -------
+    Updated input array.
+
+    Examples
+    --------
+    Given either of these equivalent expressions::
+
+      x = at(x)[1].add(2, copy=None)
+      x = at(x, 1).add(2, copy=None)
+
+    If x is a JAX array, they are the same as::
+
+      x = x.at[1].add(2)
+
+    If x is a read-only numpy array, they are the same as::
+
+      x = x.copy()
+      x[1] += 2
+
+    Otherwise, they are the same as::
+
+      x[1] += 2
+
+    Warning
+    -------
+    When you use copy=None, you should always immediately overwrite
+    the parameter array::
+
+        x = at(x, 0).set(2, copy=None)
+
+    The anti-pattern below must be avoided, as it will result in different behaviour
+    on read-only versus writeable arrays::
+
+        x = xp.asarray([0, 0, 0])
+        y = at(x, 0).set(2, copy=None)
+        z = at(x, 1).set(3, copy=None)
+
+    In the above example, ``x == [0, 0, 0]``, ``y == [2, 0, 0]`` and z == ``[0, 3, 0]``
+    when x is read-only, whereas ``x == y == z == [2, 3, 0]`` when x is writeable!
+
+    Warning
+    -------
+    The array API standard does not support integer array indices.
+    The behaviour of update methods when the index is an array of integers
+    is undefined; this is particularly true when the index contains multiple
+    occurrences of the same index, e.g. ``at(x, [0, 0]).set(2)``.
+
+    Note
+    ----
+    `sparse <https://sparse.pydata.org/>`_ is not supported by update methods yet.
+
+    See Also
+    --------
+    `jax.numpy.ndarray.at <https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.ndarray.at.html>`_
+    """
+
+    x: Array
+    idx: Index
+    __slots__: ClassVar[tuple[str, str]] = ("idx", "x")
+
+    def __init__(self, x: Array, idx: Index = _undef, /):
+        self.x = x
+        self.idx = idx
+
+    def __getitem__(self, idx: Index) -> at:
+        """Allow for the alternate syntax ``at(x)[start:stop:step]``,
+        which looks prettier than ``at(x, slice(start, stop, step))``
+        and feels more intuitive coming from the JAX documentation.
+        """
+        if self.idx is not _undef:
+            msg = "Index has already been set"
+            raise ValueError(msg)
+        self.idx = idx
+        return self
+
+    def _common(
+        self,
+        at_op: str,
+        y: Array = _undef,
+        /,
+        copy: bool | None = True,
+        xp: ModuleType | None = None,
+        _is_update: bool = True,
+        **kwargs: Untyped,
+    ) -> tuple[Untyped, None] | tuple[None, Array]:
```

Comment:
is it perhaps possible to `@overload` these cases?

Reply:
It isn't because the return type depends on a non-type test on x. And I'm definitely unwilling to explore writing a `class HasAtMethod(Protocol)` for a small internal function that is consumed exclusively 2 paragraph below.

Reply:
making this change - could you clarify how to use `_CanAt` @jorenham ?